lib/jack_range.ml

```diff
@@ -0,0 +1,59 @@
+type size = int
+
+type 'a range =
+  Range of 'a * (size -> ('a * 'a))
+
+let origin (Range (a, _))  = a
+
+let bounds sz (Range (_, f)) = f sz
+
+let singleton x =
+  Range (x, fun _ -> (x, x))
+
+let constant_from z x y =
+  Range (z, (fun _ -> (x, y)))
+
+let constant x y =
+  constant_from x x y
+
+(* let scale_linear : size -> int -> int -> int *)
+let scale_linear sz0 z n =
+  let sz = max 0 (min 99 sz0) in
+  let diff = ((n - z) * sz) / 99 in
+  z + diff
+
+(* Truncate a value so it stays within some range. *)
+let clamp x y n =
+  if x > y then
+    min x (max y n)
+  else
+    min y (max x n)
+
+let scale_exponential sz0 z0 n0 =
+  let z = float_of_int z0 in
+  let n = float_of_int n0 in
+  let sz = clamp 0 99 sz0 in
+  let diff = (((abs_float (n -. z) +. 1.0) ** (float_of_int sz /. 99.0)) -. 1.0) *. (n -. z) in
+  int_of_float @@ z +. diff
```

Comment:
Dodgy use of float / int computations here, it's likely to be wrong but for the purposes of getting something working we will just put up with it.